storage/innobase/btr/btr0btr.cc

```diff
@@ -1269,51 +1270,62 @@ btr_read_autoinc(dict_index_t* index)
 
 /** Read the last used AUTO_INCREMENT value from PAGE_ROOT_AUTO_INC,
 or fall back to MAX(auto_increment_column).
-@param[in]	table	table containing an AUTO_INCREMENT column
-@param[in]	col_no	index of the AUTO_INCREMENT column
-@return	the AUTO_INCREMENT value
-@retval	0 on error or if no AUTO_INCREMENT value was used yet */
-ib_uint64_t
-btr_read_autoinc_with_fallback(const dict_table_t* table, unsigned col_no)
+@param table          table containing an AUTO_INCREMENT column
+@param col_no         index of the AUTO_INCREMENT column
+@param mysql_version  TABLE_SHARE::mysql_version
+@param max            the maximum value of the AUTO_INCREMENT column
+@return the AUTO_INCREMENT value
+@retval 0 on error or if no AUTO_INCREMENT value was used yet */
+uint64_t btr_read_autoinc_with_fallback(const dict_table_t *table,
+                                        unsigned col_no, ulong mysql_version,
+                                        uint64_t max)
 {
-	ut_ad(table->persistent_autoinc);
-	ut_ad(!table->is_temporary());
+  ut_ad(table->persistent_autoinc);
+  ut_ad(!table->is_temporary());
 
-	dict_index_t*	index = dict_table_get_first_index(table);
+  uint64_t autoinc= 0, max_autoinc= 0;
+  dict_index_t *index= dict_table_get_first_index(table);
+  mtr_t mtr;
+  mtr.start();
 
-	if (index == NULL) {
-		return 0;
-	}
+  if (buf_block_t *block=
+      buf_page_get(page_id_t(table->space_id, index->page),
+                   table->space->zip_size(), RW_SX_LATCH, &mtr))
+  {
+    autoinc= page_get_autoinc(block->frame);
 
-	mtr_t		mtr;
-	mtr.start();
-	buf_block_t*	block = buf_page_get(
-		page_id_t(index->table->space_id, index->page),
-		index->table->space->zip_size(),
-		RW_S_LATCH, &mtr);
+    if (autoinc > max || mysql_version == 0 || mysql_version < 100210)
+    {
+      /* Read MAX(autoinc_col), in case this table had originally been
+      created before MariaDB 10.2.4 introduced persistent AUTO_INCREMENT
+      and MariaDB 10.2.10 fixed MDEV-12123, and there could be a garbage
```

Comment:
Theoretically, we could remove the second `if` condition in the code that you quoted, to always reset `PAGE_MAX_TRX_ID` in clustered index pages, including when it corresponds to `PAGE_ROOT_AUTO_INC`.

But, that would make us lose the ability to preserve a persistent `AUTO_INCREMENT` value on `ALTER TABLE…IMPORT TABLESPACE` when not using a `.cfg` file. In [MDEV-11658](https://jira.mariadb.org/browse/MDEV-11658) I would want to get rid of `.cfg` files.

Before [MDEV-26137](https://jira.mariadb.org/browse/MDEV-26137) (9b431d714fdf030e372ed6eeaec0619dcffdde6a) we could not use the original `.frm` file when importing InnoDB data files. Only by using the original `.frm` file it would be possible to augment this condition so that we would preserve legitimate persistent `AUTO_INCREMENT` values.

I think that this suggestion would best be implemented near the end of `row_import_for_mysql()` where we can conveniently access the entire index tree via the buffer pool. At that point, we could invoke this function `btr_read_autoinc_with_fallback()` with appropriate parameters. This should be doable after #3037 has been merged.

Reply:
Addressed in efb8c05d352fbf27849866ad8ba8d2c5f822c03f.